### PR TITLE
Objective function support

### PR DESCRIPTION
**Pull request by javdrher**
Noticed we actually have zero documentation on how to write an objective function. Fixed this, and added two decorators which accomplish two highly recurring tasks (applying a function in batch, and splitting columns to function parameters). 

This branch also introduced a tiny library for parameterized unittesting. Its not a main dependency, only for testing.

### CODE REVIEW

File: GPflowOpt/objective.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+        responses = (fun(x) for x in np.atleast_2d(X))
+        sep = tuple(zip(*(r if isinstance(r, tuple) else (r,) for r in responses)))
+        f = np.vstack(sep[0])
+        if len(sep) == 1:
+            return f
+
+        # for each point, the gradient is either (d,) or (d, p) shaped.
+        g_stacked = np.stack((r for r in sep[1]), axis=0)  # n x d or n x d x p
+        # Get rid of last dim = 1 in case p = 1
+        g = np.squeeze(g_stacked, axis=2) if len(g_stacked.shape) == 3 and g_stacked.shape[2] == 1 else g_stacked
```
Comment (icouckuy) [inline GPflowOpt/objective.py]: why do we have to squeeze if p =1 ?

File: GPflowOpt/objective.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+       dimensionality n x p and optionally a gradient of size n x d x p (or n x d if p == 1)
+    :return: a function wrapper which splits a given input ndarray into its columns to call fun.
+    """
+    @wraps(fun)
+    def args_wrapper(X):
+        X = np.atleast_2d(X)
+        return fun(*X.T)
+
+    return args_wrapper
+
```
Comment (icouckuy) [inline GPflowOpt/objective.py]: what was the reason again a to_kwars() accepting the domain object was not possible?

File: GPflowOpt/objective.py (hunk 1 above)
Comment (javdrher) [inline GPflowOpt/objective.py]: Currently we don't have any gradient-based optimizers for multiple objectives, so the n x p x d case doesn't occur currently. Instead, right now n x d matrices are returned.

File: GPflowOpt/objective.py (hunk 2 above)
Comment (javdrher) [inline GPflowOpt/objective.py]: hm, we could actually add that one as well but this one seems to have a reason of existence?

Comment (icouckuy) [inline GPflowOpt/objective.py]: I think both to_args and to_kwargs will be useful. to_kwargs should work with batch_apply too.

At the moment I am using a lambda function to convert the numpy matrix to kwargs. This is useful as I do not need to change my objective function when I change the dimensionality (the kwargs have default values).

If we go for the idea of setting parameters fixed, similarly to GPflow, I can also use a to_args (but kwargs  will also be useful I guess).

**Pull request merged**